File: erpnext/accounts/page/accounts_dashboard/accounts_dashboard.py
import frappe
from datetime import datetime
from frappe.utils import add_to_date, date_diff, getdate, nowdate, get_last_day, formatdate
from erpnext.accounts.dashboard_chart_source.account_balance_timeline.account_balance_timeline import build_result, get_dates_from_timegrain, get_gl_entries
from erpnext.accounts.report.profit_and_loss_statement.profit_and_loss_statement import execute
from frappe.core.page.dashboard.dashboard import get_from_date_from_timespan
from frappe.desk.doctype.dashboard_chart.dashboard_chart import get_period_ending

@frappe.whitelist()
def get_bank_balances():
    #get the default company
    company = frappe.defaults.get_user_default("company")
    if not company:
        frappe.throw(_("No default company found. Set default company in Session Defaults"))
    currency = frappe.db.get_value("Company", company, 'default_currency')
    currency = frappe.db.get_value("Currency", currency, 'symbol')
    #get the bank account of company
    abbr = frappe.db.get_value("Company", company, 'abbr')
    account = 'Bank Accounts - ' + abbr
    #get non-group descendents of bank accounts
    lft, rgt = frappe.db.get_value('Account', account, ['lft', 'rgt'])
    accounts = [d["name"] for d in frappe.db.get_list('Account', {"lft": [">", lft], "rgt": ["<", rgt], "is_group": 0},
		"name", order_by='lft desc', limit_page_length=None, ignore_permissions=False)]
    to_date = nowdate()
    if not accounts:
        return 'No accounts found'
    dates = [to_date]
    accounts_balances = []
    #loop through all accounts and get their balances
    for a in accounts:
        gl_entries = frappe.db.get_all('GL Entry',
        fields = ['posting_date', 'debit', 'credit'],
        filters = [
			dict(posting_date = ('<=', to_date)),
			dict(account = ('=', a))
		],
		order_by = 'posting_date asc')
        result = build_result(a, dates, gl_entries)
        accounts_balances.append(result[0][1])
    balances = dict(zip(accounts, accounts_balances))
    bal = {
        "balances": balances,
        "currency": currency
    }
    return bal

@frappe.whitelist()
def get_fiscal_years():
    fiscal_years = []
    fiscals = frappe.get_list('Fiscal Year', fields=['name'])
    for f in fiscals:
        fiscal_years.append(f.name)
    return fiscal_years

@frappe.whitelist()
def get_invoices(invoice_type, timespan, company):
    currency = frappe.db.get_value("Company", company, 'default_currency')
    currency = frappe.db.get_value("Currency", currency, 'symbol')
    to_date = nowdate()
    from_date = get_from_date_from_timespan(to_date, timespan)
    if invoice_type == 'Sales Invoice':
        frappe.cache().hset('i_span', frappe.session.user, timespan)
    if invoice_type == 'Purchase Invoice':
        frappe.cache().hset('ip_span', frappe.session.user, timespan)
    paid_invoices = frappe.get_list(invoice_type, filters=[["company", "=", company], ["status", "=", "Paid"], ["posting_date", ">=", from_date], ["posting_date", "<=", to_date]], fields=['status', 'grand_total'])
    unpaid_invoices = frappe.get_list(invoice_type, filters=[["company", "=", company], ["status", "=", "Unpaid"], ["posting_date", ">=", from_date], ["posting_date", "<=", to_date]], fields=['status', 'grand_total'])
    overdue_invoices = frappe.get_list(invoice_type, filters=[["company", "=", company], ["status", "=", "Overdue"], ["posting_date", ">=", from_date], ["posting_date", "<=", to_date]], fields=['status', 'grand_total'])
    paid_total = sum(p['grand_total'] for p in paid_invoices)
    unpaid_total = sum(u['grand_total'] for u in unpaid_invoices)
    overdue_total = sum(o['grand_total'] for o in overdue_invoices)
    chart = {
        "data": {
        'labels': ['Paid Invoices', 'Unpaid Invoices', 'Overdue'],
        'datasets': [{
            'label': '# of Votes',
            'data': [len(paid_invoices), len(unpaid_invoices), len(overdue_invoices)],
            'backgroundColor': [
                '#4f1ebb',
                '#aaa3d0',
                '#ff5858'
            ],
        }]
    }
    }
    chart["paid_total"] = paid_total
    chart["unpaid_total"] = unpaid_total
    chart["overdue_total"] = overdue_total
    chart["currency"] = currency
    return chart

@frappe.whitelist()
def get(chart = None):
    company = frappe.defaults.get_user_default("company")
    abbr = frappe.db.get_value("Company", company, 'abbr')
    chart = frappe._dict(frappe.parse_json(chart))
    timespan = chart.timespan
    timegrain = chart.time_interval
    frappe.cache().hset('sa_span', frappe.session.user, timespan)
    frappe.cache().hset('sa_interval', frappe.session.user, timegrain)
    account = 'Sales - ' + abbr
    to_date = nowdate()
    from_date = get_from_date_from_timespan(to_date, timespan)
    dates = get_dates_from_timegrain(from_date, to_date, timegrain)
    gl_entries = get_gl_entries(account, get_period_ending(to_date, timegrain))
    result = build_result(account, dates, gl_entries)
    return {
        "labels": [formatdate(r[0].strftime('%Y-%m-%d')) for r in result],
        "datasets": [{
            "name": account,
            "values": [r[1] for r in result]
            }]
    }

@frappe.whitelist()
def cache_dashboard_values():
    frappe.cache().hset("i_span", frappe.session.user, "Last Year")
    frappe.cache().hset("ip_span", frappe.session.user, "Last Year")
    frappe.cache().hset('pl_span', frappe.session.user, 'Quarterly')
    frappe.cache().hset('pl_chart', frappe.session.user, 'line')
    frappe.cache().hset('pl_from', frappe.session.user, frappe.defaults.get_user_default("fiscal_year"))
    frappe.cache().hset('pl_to', frappe.session.user, frappe.defaults.get_user_default("fiscal_year"))
    frappe.cache().hset('pl_navigable', frappe.session.user, 1)
    frappe.cache().hset('pl_values', frappe.session.user, 1)
    frappe.cache().hset('pl_fill', frappe.session.user, 1)
    frappe.cache().hset('pl_dots', frappe.session.user, 0)
    frappe.cache().hset('pl_line', frappe.session.user, 0)
    frappe.cache().hset('pl_heatline', frappe.session.user, 0)
    frappe.cache().hset('pl_stack', frappe.session.user, 0)
    frappe.cache().hset('cf_chart', frappe.session.user, 'bar')
    frappe.cache().hset('cf_span', frappe.session.user, 'Yearly')
    frappe.cache().hset('cf_from', frappe.session.user, frappe.defaults.get_user_default("fiscal_year"))
    frappe.cache().hset('cf_to', frappe.session.user, frappe.defaults.get_user_default("fiscal_year"))
    frappe.cache().hset('cf_navigable', frappe.session.user, 0)
    frappe.cache().hset('cf_values', frappe.session.user, 1)
    frappe.cache().hset('cf_fill', frappe.session.user, 1)
    frappe.cache().hset('cf_dots', frappe.session.user, 0)
    frappe.cache().hset('cf_line', frappe.session.user, 0)
    frappe.cache().hset('cf_heatline', frappe.session.user, 0)
    frappe.cache().hset('cf_stack', frappe.session.user, 0)
    frappe.cache().hset('sa_span', frappe.session.user, 'Last Year')
    frappe.cache().hset('sa_interval', frappe.session.user, 'Monthly')
    return True
 

@frappe.whitelist()
def get_cache_values():
    invoice_values, p_values, c_values, sales_values = {}, {}, {}, {}
    invoice = ['i_span', 'ip_span']
    p_chart = [ 'pl_chart', 'pl_span', 'pl_from', 'pl_to',
     'pl_navigable', 'pl_values', 'pl_fill', 'pl_dots', 'pl_line',
     'pl_heatline', 'pl_stack']
    c_chart = ['cf_chart', 'cf_span', 'cf_from', 'cf_to',
     'cf_navigable', 'cf_values', 'cf_fill', 'cf_dots', 'cf_line', 'cf_heatline',
     'cf_stack']
    sales = ['sa_span', 'sa_interval']
    for i in invoice:
        v = frappe.cache().hget(i, frappe.session.user)
        invoice_values[i] = v
    print(invoice_values)
    for p in p_chart:
        v = frappe.cache().hget(p, frappe.session.user)
        p_values[p] = v
    for c in c_chart:
        v = frappe.cache().hget(c, frappe.session.user)
        c_values[c] = v
    for s in sales:
        v = frappe.cache().hget(s, frappe.session.user)
        sales_values[s] = v
    cached_values = {
        'invoice_values': invoice_values,
        'p_values': p_values,
        'c_values': c_values,
        'sales_values': sales_values
    }
    return cached_values

@frappe.whitelist()
def set_custom_filters(filtername, filtervalue):
    if filtervalue == 'false':
        filtervalue = 0
    if filtervalue == 'true':
        filtervalue = 1
    frappe.cache().hset(filtername, frappe.session.user, filtervalue)

File: erpnext/accounts/page/accounts_dashboard/accounts_dashboard.js
frappe.pages['accounts-dashboard'].on_page_load = function(wrapper) {

	frappe.ui.make_app_page({
		parent: wrapper,
		title: __('Accounts Dashboard'),
	});

	let acc = new AccountsDashboard(wrapper);
	$(wrapper).bind('show', ()=> {
		acc.show();
	});
};

class AccountsDashboard {

	constructor(wrapper) {
		this.wrapper = $(wrapper);
		this.page = wrapper.page;
		this.main_section = this.wrapper.find('.layout-main-section');
		$('[data-toggle="tooltip"]').tooltip();
	}

	show() {
		this.$user_search_button = this.page.set_secondary_action('Set to Default', () => {
			frappe.xcall('erpnext.accounts.page.accounts_dashboard.accounts_dashboard.cache_dashboard_values', {
				async: false,
			}).then(chart => {
				this.show();
			});
		});
		this.main_section.empty().append(frappe.render_template('accounts_dashboard'));
		this.bank_balances = this.main_section.find('.bank-balances');
		this.sales_doughnut = this.main_section.find('.sales-doughnut');
		this.purchase_doughnut = this.main_section.find('.purchase-doughnut');
		this.get_cached_filters();
	}

	get_cached_filters() {
		  return frappe.call({
			method: 'erpnext.accounts.page.accounts_dashboard.accounts_dashboard.get_cache_values',
			async: false,
			callback: r => {
				console.log(r)
				if(!r.message.sales_values.sa_span){
					this.set_defaults();
					console.log('yes')
				}
				else{
					this.s_filters = {
						'timespan': r.message.sales_values.sa_span,
						'time_interval': r.message.sales_values.sa_interval,
					}
					this.invoice_filter = r.message.invoice_values.i_span;
					this.pl_filters = {
						'company': frappe.defaults.get_user_default("company"),
						'to_fiscal_year': r.message.p_values.pl_to,
						'from_fiscal_year': r.message.p_values.pl_from,
						'periodicity': r.message.p_values.pl_span
					};
					this.pl_customs = {
						type: r.message.p_values.pl_chart,
						isNavigable: r.message.p_values.pl_navigable, 
						valuesOverPoints: r.message.p_values.pl_values,
						regionFill: r.message.p_values.pl_fill,
						hideDots: r.message.p_values.pl_dots,
						hideLine: r.message.p_values.pl_line,
						heatline: r.message.p_values.pl_heatline,
						stacked: r.message.p_values.pl_stack
					}
					this.cf_filters = {
						'company': frappe.defaults.get_user_default("company"),
						'to_fiscal_year': r.message.c_values.cf_to,
						'from_fiscal_year': r.message.c_values.cf_from,
						'periodicity': r.message.c_values.cf_span
					};
					this.cf_customs = {
						type: r.message.c_values.cf_chart,
						isNavigable: r.message.c_values.cf_navigable, 
						valuesOverPoints: r.message.c_values.cf_values,
						regionFill: r.message.c_values.cf_fill,
						hideDots: r.message.c_values.cf_dots,
						hideLine: r.message.c_values.cf_line,
						heatline: r.message.c_values.cf_heatline,
						stacked: r.message.c_values.cf_stack
					}
					this.render_profit_loss();
					if (frappe.user.has_role('Administrator') || frappe.user.has_role('System Manager') || frappe.user.has_role('Accounts Manager') || frappe.user.has_role('Accounts User')){
					this.render_bank_balances();
					this.main_section.find('.invoice-c').removeClass('hide');
					this.get_sales_invoices();
					this.get_purchase_invoices();
					this.create_invoice_filters();
					this.main_section.find('.cf-chart-container').removeClass('hide');
					this.render_cash_flow();
					this.main_section.find('.sales-chart-container').removeClass('hide');
					this.render_sales();
					}
			}
			},
		  })
	}

	
	set_defaults() {
		frappe.xcall('erpnext.accounts.page.accounts_dashboard.accounts_dashboard.cache_dashboard_values', {
			async: false,
		}).then(chart => {
			this.get_cached_filters();
		});
	}

	set_custom_filters(filter_name, filter_value) {
		console.log(filter_value)
		frappe.xcall('erpnext.accounts.page.accounts_dashboard.accounts_dashboard.set_custom_filters', {
			filtername: filter_name,
			filtervalue: filter_value
		}).then(chart => {
		});
	}

	get_sales() {
		frappe.xcall('erpnext.accounts.page.accounts_dashboard.accounts_dashboard.get', {
			chart: this.s_filters
		}).then(chart => {
			this.sales_chart.update(chart)
		});
	}

	render_sales() {
		this.sales_chart = new frappe.Chart('.sales-chart', {
			type: 'bar', //customizable
			height: 300,
			colors: ['#4f1ebb'],
			data: {
				labels: [],
	  
			datasets: [
			  {
				name: 'Sales',
				values: []
			  }
			]
			},
			tooltipOptions: {
						formatTooltipX: d => d,
						formatTooltipY: d => frappe.format(d, {
							fieldtype: "Currency",
						})
					}
		});
		this.get_sales();
		this.create_sales_chart_filters();
	}

	create_sales_chart_filters(){
		let filters = [
			{
				label: this.s_filters.timespan,
				options: ['Last Year', 'Last Quarter', 'Last Month'],
				action: (selected_item) => {
					this.s_filters.timespan = selected_item;
					this.get_sales();
				}
			},
			{
				label: this.s_filters.time_interval,
				options: ['Quarterly', 'Monthly', 'Weekly', 'Daily'],
				action: (selected_item) => {
					this.s_filters.time_interval = selected_item;
					this.get_sales();
				}
			},
		];
		this.render_chart_filters(filters, '.sales-chart-container', 1);
	}

	render_bank_balances() {
		frappe.xcall('erpnext.accounts.page.accounts_dashboard.accounts_dashboard.get_bank_balances', {
		}).then((r) => {
			this.bank_balances.empty().append(frappe.render_template('bank_balances', {
				abc: r.balances,
				currency: r.currency
			}));
		});
	}

	get_fiscal_years(){
		return frappe.call({
		  method: 'erpnext.accounts.page.accounts_dashboard.accounts_dashboard.get_fiscal_years',
		  async: false,
		  callback: r => {
			this.fiscals = r.message
		  },
		});
}

	render_profit_loss() {
		this.make_pl_chart();
		this.create_pl_chart_filters();
		this.create_pl_chart_customs();
	}

	make_pl_chart(){
		console.log(this.pl_customs.isNavigable)
		this.pl_chart = new frappe.Chart('.pl-chart', {
			type: this.pl_customs.type, //customizable
			height: 300,
			colors: ['#0e6333', '#ff5858', '#4f1ebb'],
			isNavigable: this.pl_customs.isNavigable, //customizable
			valuesOverPoints: this.pl_customs.valuesOverPoints, //customizable
			lineOptions: {
				regionFill: this.pl_customs.regionFill, //customizable
				hideDots: this.pl_customs.hideDots, //customizable
				hideLine: this.pl_customs.hideLine, //customizable
				heatline: this.pl_customs.heatline //customizable
			},
			barOptions: {
				stacked: this.pl_customs.stacked, //customizable
			},
			data: {
				labels: [],
	  
			datasets: [
			  {
				name: 'Income',
				values: []
			  },
			  {
				name: 'Expenses',
				values: []
			  },
			  {
				name: 'Net Profit/Loss',
				values: []
			  }
			],
			},
			tooltipOptions: {
						formatTooltipX: d => d,
						formatTooltipY: d => frappe.format(d, {
							fieldtype: "Currency",
						})
					}
		});
		this.update_pl_chart_data();
	}


	update_pl_chart_data() {
		frappe.xcall('erpnext.accounts.report.profit_and_loss_statement.profit_and_loss_statement.execute', {
			filters: this.pl_filters,
			dashboard: true,
		}).then(chart => {
			this.pl_chart.update(chart.data);
		});
	}

	create_pl_chart_filters() {
		this.fiscals = []
		this.get_fiscal_years();
		let filters = [
			{
				label: this.pl_customs.type,
				options: ['line', 'bar'],
				action: (selected_item) => {
					this.pl_customs.type = selected_item
					this.make_pl_chart();
					this.set_custom_filters('pl_chart', this.pl_customs.type)
				}
			},
			{
				label: this.pl_filters.periodicity,
				options: ['Yearly', 'Half-Yearly', 'Quarterly', 'Monthly'],
				action: (selected_item) => {
					this.pl_filters.periodicity = selected_item;
					this.update_pl_chart_data();
				}
			},
			{
				label: this.pl_filters.from_fiscal_year,
				title: "From Fiscal Year",
				options: this.fiscals,
				action: (selected_item) => {
					this.pl_filters.from_fiscal_year = selected_item;
					this.update_pl_chart_data();
				}
			},
			{
				label: this.pl_filters.to_fiscal_year,
				title: "To Fiscal Year",
				options: this.fiscals,
				action: (selected_item) => {
					this.pl_filters.to_fiscal_year = selected_item;
					this.update_pl_chart_data();
				}
			},
		];
		this.render_chart_filters(filters, '.pl-chart-container', 1, 'pl_chart');
	}

	render_cash_flow() {
		this.make_cf_chart();
		this.create_cf_chart_filters();
		this.create_cf_chart_customs();
	}


	make_cf_chart(){
		this.cf_chart = new frappe.Chart('.cf-chart', {
			type: this.cf_customs.type, //customizable
			height: 200,
			colors: ['#0e6333', '#ff5858', '#4f1ebb'],
			isNavigable: this.cf_customs.isNavigable, //customizable
			valuesOverPoints: this.cf_customs.valuesOverPoints, //customizable
			lineOptions: {
				regionFill: this.cf_customs.regionFill, //customizable
				hideDots: this.cf_customs.hideDots, //customizable
				hideLine: this.cf_customs.hideLine, //customizable
				heatline: this.cf_customs.heatline //customizable
			},
			barOptions: {
				stacked: this.cf_customs.stacked, //customizable
			},
			data: {
				labels: [],
	  
				datasets: [
					{
					  name: 'Net Cash from Operations',
					  values: []
					},
					{
					  name: 'Net Cash from Investing',
					  values: []
					},
					{
					  name: 'Net Cash from Financing',
					  values: []
					},
				  ],
			},
			tooltipOptions: {
						formatTooltipX: d => d,
						formatTooltipY: d => frappe.format(d, {
							fieldtype: "Currency",
						})
					}
		});
		this.update_cf_chart_data();
	}

	update_cf_chart_data() {
		frappe.xcall('erpnext.accounts.report.cash_flow.cash_flow.execute', {
			filters: this.cf_filters,
			dashboard: true,
		}).then(chart => {
			this.cf_chart.update(chart.data);
		});
	}

	create_cf_chart_filters() {
		this.fiscals = []
		this.get_fiscal_years();
		let filters = [
			{
				label: this.cf_customs.type,
				options: ['bar', 'line'],
				action: (selected_item) => {
					this.cf_customs.type = selected_item
					this.make_cf_chart();
					this.set_custom_filters('cf_chart', this.cf_customs.type)
				}
			},
			{
				label: this.cf_filters.periodicity,
				options: ['Yearly', 'Half-Yearly', 'Quarterly', 'Monthly'],
				action: (selected_item) => {
					this.cf_filters.periodicity = selected_item;
					this.update_cf_chart_data();
				}
			},
			{
				label: this.cf_filters.from_fiscal_year,
				title: "From Fiscal Year",
				options: this.fiscals,
				action: (selected_item) => {
					this.cf_filters.from_fiscal_year = selected_item;
					this.update_cf_chart_data();
				}
			},
			{
				label: this.cf_filters.to_fiscal_year,
				title: "To Fiscal Year",
				options: this.fiscals,
				action: (selected_item) => {
					this.cf_filters.to_fiscal_year = selected_item;
					this.update_cf_chart_data();
				}
			},
		];
		this.render_chart_filters(filters, '.cf-chart-container', 1, 'cf_chart');
	}

	render_chart_filters(filters, container, append, chart_name) {
		let export_btn = `<button class="btn btn-default btn-xs">
		<span class="filter-label">Export Chart</span>
	</button>`;
		let $export_chart = $(export_btn);
		if(chart_name == 'pl_chart'){
		$export_chart.appendTo(this.wrapper.find(container));
		$export_chart.on('click', (e) => {
				this.pl_chart.export();
			});
	}
	else if(chart_name == 'cf_chart'){
		$export_chart.appendTo(this.wrapper.find(container));
		$export_chart.on('click', (e) => {
				this.cf_chart.export();
			});
	}
		filters.forEach(filter => {
			let chart_filter_html = `<div class="chart-filter pull-right">
			<a class="dropdown-toggle" data-toggle="dropdown" aria-haspopup="true" aria-expanded="false">
					<button class="btn btn-default btn-xs" data-toggle="tooltip" data-placement="top" title="${filter.title}">
						<span class="filter-label">${filter.label}</span>
						<span class="caret"></span>
					</button>
				</a>`;
			let options_html;

			if (filter.fieldnames) {
				options_html = filter.options.map((option, i) =>
					`<li><a data-fieldname = "${filter.fieldnames[i]}">${option}</a></li>`).join('');
			} else {
				options_html = filter.options.map( option => `<li><a>${option}</a></li>`).join('');
			}

			let dropdown_html = chart_filter_html + `<ul class="dropdown-menu">${options_html}</ul></div>`;
			let $chart_filter = $(dropdown_html);

			if (append) {
				$chart_filter.prependTo(this.wrapper.find(container));
			} else $chart_filter.appendTo(this.wrapper.find(container));

			$chart_filter.find('.dropdown-menu').on('click', 'li a', (e) => {
				let $el = $(e.currentTarget);
				let fieldname;
				if ($el.attr('data-fieldname')) {
					fieldname = $el.attr('data-fieldname');
				}
				let selected_item = $el.text();
				$el.parents('.chart-filter').find('.filter-label').text(selected_item);
				filter.action(selected_item, fieldname);
			});
		});

	}

	create_cf_chart_customs() {
		let customs = [
			{
				label: 'Navigable',
				value: this.cf_customs.isNavigable,
				action: (selected_item) => {
					this.cf_customs.isNavigable = selected_item;
					this.make_cf_chart();
					this.set_custom_filters('cf_navigable', this.cf_customs.isNavigable)
				}
			},
			{
				label: 'Values over points',
				value: this.cf_customs.valuesOverPoints,
				action: (selected_item) => {
					this.cf_customs.valuesOverPoints = selected_item;
					this.make_cf_chart();
					this.set_custom_filters('cf_values', this.cf_customs.valuesOverPoints)
				}
			},
			{
				label: 'Fill region',
				value: this.cf_customs.regionFill,
				action: (selected_item) => {
					this.cf_customs.regionFill = selected_item;
					this.make_cf_chart();
					this.set_custom_filters('cf_fill', this.cf_customs.regionFill)
				}
			},
			{
				label: 'Hide Dots',
				value: this.cf_customs.hideDots,
				action: (selected_item) => {
					this.cf_customs.hideDots = selected_item;
					this.make_cf_chart();
					this.set_custom_filters('cf_dots', this.cf_customs.hideDots)
				}
			},
			{
				label: 'Hide Line',
				value: this.cf_customs.hideLine,
				action: (selected_item) => {
					this.cf_customs.hideLine = selected_item;
					this.make_cf_chart();
					this.set_custom_filters('cf_line', this.cf_customs.hideLine)
				}
			},
			{
				label: 'Heatline',
				value: this.cf_customs.heatline,
				action: (selected_item) => {
					this.cf_customs.heatline = selected_item;
					this.make_cf_chart();
					this.set_custom_filters('cf_heatline', this.cf_customs.heatline)
				}
			},
			{
				label: 'Stacked',
				value: this.cf_customs.stacked,
				action: (selected_item) => {
					this.cf_customs.stacked = selected_item;
					this.make_cf_chart();
					this.set_custom_filters('cf_stack', this.cf_customs.stacked)
				}
			},
		];
		this.render_chart_customs(customs, '.cf-chart-container', 1);
	}

	create_pl_chart_customs() {
		console.log(this.pl_customs.isNavigable)
		let customs = [
			{
				label: 'Navigable',
				value: this.pl_customs.isNavigable,
				action: (selected_item) => {
					this.pl_customs.isNavigable = selected_item;
					this.make_pl_chart();
					this.set_custom_filters('pl_navigable', this.pl_customs.isNavigable)
				}
			},
			{
				label: 'Values over points',
				value: this.pl_customs.valuesOverPoints,
				action: (selected_item) => {
					this.pl_customs.valuesOverPoints = selected_item;
					this.make_pl_chart();
					this.set_custom_filters('pl_values', this.pl_customs.valuesOverPoints)
				}
			},
			{
				label: 'Fill region',
				value: this.pl_customs.regionFill,
				action: (selected_item) => {
					this.pl_customs.regionFill = selected_item;
					this.make_pl_chart();
					this.set_custom_filters('pl_fill', this.pl_customs.regionFill)
				}
			},
			{
				label: 'Hide Dots',
				value: this.pl_customs.hideDots,
				action: (selected_item) => {
					this.pl_customs.hideDots = selected_item;
					this.make_pl_chart();
					this.set_custom_filters('pl_dots', this.pl_customs.hideDots)
				}
			},
			{
				label: 'Hide Line',
				value: this.pl_customs.hideLine,
				action: (selected_item) => {
					this.pl_customs.hideLine = selected_item;
					this.make_pl_chart();
					this.set_custom_filters('pl_line', this.pl_customs.hideLine)
				}
			},
			{
				label: 'Heatline',
				value: this.pl_customs.heatline,
				action: (selected_item) => {
					this.pl_customs.heatline = selected_item;
					this.make_pl_chart();
					this.set_custom_filters('pl_heatline', this.pl_customs.heatline)
				}
			},
			{
				label: 'Stacked',
				value: this.pl_customs.stacked,
				action: (selected_item) => {
					this.pl_customs.stacked = selected_item;
					this.make_pl_chart();
					this.set_custom_filters('pl_stack', this.pl_customs.stacked)
				}
			},
		];
		this.render_chart_customs(customs, '.pl-chart-container', 1);
	}

	render_chart_customs(filters, container, append) {
		filters.forEach(filter => {
			let chart_filter_html = `<div class="checkbox inline-checkboxes">
			<label><input type="checkbox">${filter.label}</label>
		  </div>`;
			let $chart_filter = $(chart_filter_html);
			  $chart_filter.find('input[type="checkbox"]').prop('checked', filter.value);
			  $chart_filter.appendTo(this.wrapper.find(container));
			$chart_filter.find("input[type='checkbox']").on('change', (e) => {
				let $el = $(e.currentTarget);
				let fieldname;
				if ($el.attr('data-fieldname')) {
					fieldname = $el.attr('data-fieldname');
				}
				let selected_item = $el[0].checked;
				filter.action(selected_item, fieldname);
			});
		});

	}

	create_sales_invoice_chart(data) {
var ctx = document.getElementById('myChart').getContext('2d');
var myChart = new Chart(ctx, {
    type: 'doughnut',
    data: data,
    options: {
		cutoutPercentage: 70,
		responsive: false,
		legend: {
			display: false
		},
    }
});
	}

	create_purchase_invoice_chart(data) {
var ctx = document.getElementById('purchaseI').getContext('2d');
var myChart = new Chart(ctx, {
    type: 'doughnut',
    data: data,
    options: {
		cutoutPercentage: 70,
		responsive: false,
		legend: {
			display: false
		},
    }
});
	}

	get_sales_invoices() {
		frappe.xcall('erpnext.accounts.page.accounts_dashboard.accounts_dashboard.get_invoices', {
			invoice_type: 'Sales Invoice',
			timespan: this.invoice_filter,
			company: frappe.defaults.get_user_default("company")
		}).then(r => {
			this.sales_doughnut.empty().append(frappe.render_template('sales', {
				type: 'Sales',
				paid_total: r.paid_total,
				unpaid_total: r.unpaid_total,
				overdue_total: r.overdue_total,
				currency: r.currency
			}));
			this.create_sales_invoice_chart(r.data);
		});
	}

	get_purchase_invoices() {
		frappe.xcall('erpnext.accounts.page.accounts_dashboard.accounts_dashboard.get_invoices', {
			invoice_type: 'Purchase Invoice',
			timespan: this.invoice_filter,
			company: frappe.defaults.get_user_default("company")
		}).then(r => {
			this.purchase_doughnut.empty().append(frappe.render_template('purchase', {
				type: 'Purchases',
				paid_total: r.paid_total,
				unpaid_total: r.unpaid_total,
				overdue_total: r.overdue_total,
				currency: r.currency
			}));
			this.create_purchase_invoice_chart(r.data);
		});
	}

	create_invoice_filters() {
		let filters = [
			{
				label: this.invoice_filter,
				options: ['Last Year', 'Last Month', 'Last Quarter', 'Last Week'],
				action: (selected_item) => {
					this.invoice_filter = selected_item;
					this.get_sales_invoices();
					this.get_purchase_invoices();
				}
			}
		];
		this.render_invoice_filters(filters, '.invoice-chart-container', 1, 'none');
	}

	render_invoice_filters(filters, container, append, chart_name) {
		filters.forEach(filter => {
			let chart_filter_html = `<div class="chart-filter">
			<a class="dropdown-toggle" data-toggle="dropdown" aria-haspopup="true" aria-expanded="false">
					<button class="btn btn-default btn-xs" data-toggle="tooltip" data-placement="top" title="${filter.title}">
						<span class="filter-label">${filter.label}</span>
						<span class="caret"></span>
					</button>
				</a>`;
			let options_html;

			if (filter.fieldnames) {
				options_html = filter.options.map((option, i) =>
					`<li><a data-fieldname = "${filter.fieldnames[i]}">${option}</a></li>`).join('');
			} else {
				options_html = filter.options.map( option => `<li><a>${option}</a></li>`).join('');
			}

			let dropdown_html = chart_filter_html + `<ul class="dropdown-menu">${options_html}</ul></div>`;
			let $chart_filter = $(dropdown_html);

			if (append) {
				$chart_filter.prependTo(this.wrapper.find(container));
			} else $chart_filter.appendTo(this.wrapper.find(container));

			$chart_filter.find('.dropdown-menu').on('click', 'li a', (e) => {
				let $el = $(e.currentTarget);
				let fieldname;
				if ($el.attr('data-fieldname')) {
					fieldname = $el.attr('data-fieldname');
				}
				let selected_item = $el.text();
				$el.parents('.chart-filter').find('.filter-label').text(selected_item);
				filter.action(selected_item, fieldname);
			});
		});

	}
}
